### ISSUE TITLE
The app is not compatible 

### ISSUE THREAD

**Issue by JamesBlack2** [dotnet/wpf#107]
我使用VS2019新建了一个测试用的.net core的控制台应用程序，但是当我把这程序发布后生成的文件复制到Win7 service 1,7061的 32位系统想远行时，却提示文件版本与系统不兼容，就是将其设置兼容性再跑时，也不行。是否VS2019预览版并没有完全将.net core的运行打包在一起？

下面是app的项目配置。
《Project Sdk="Microsoft.NET.Sdk">
 《 PropertyGroup>
   《 OutputType>Exe《/OutputType>
   《 TargetFramework>netcoreapp2.2《/TargetFramework>
  《/PropertyGroup>
《/Project>

**Comment by lindexi** [comment]
You need to install the dotnet core runtime in win7.

你需要在 Win7 的系统上安装 运行时。

Or you can publish the dotnet core application to win7 x86.

或者将 dotnet core 程序发布为 Win7 x86 然后复制到 win7 运行

**Comment by JamesBlack2** [comment]
有两个问题：
1）
2>正在还原 H:\Test\WpfApp6\WpfApp6\WpfApp6.csproj 的包...

这个步骤不动了，我是发布到本地的目录上的，这个应该不需要这么久的时间吧。这个步骤在我这至少花了好几分钟。不过后面几次很快，这个我测试了几个工程，第一次都需要好几分钟时间。

2）在那个控制台应用程序，通过发布独立的，是可以正常跑，但是如果是WPF的话，则一打开就会崩溃跑不起来。

WPF的项目配置如下：
《Project Sdk="Microsoft.NET.Sdk.WindowsDesktop">
  《PropertyGroup>
    《OutputType>WinExe</OutputType>
    《TargetFramework>netcoreapp3.0</TargetFramework>
   《UseWPF>true</UseWPF>
    《AssemblyName>MyCore3App《/AssemblyName>
   《Platforms>AnyCPU;x64;x86《/Platforms>
  《/PropertyGroup>
《/Project>

**Comment by karelz** [comment]
In general, we take feedback and questions only in English. If it still needs a response, can someone please post English translations? (automated are mostly fine)
We can then update the original posts with appropriate text.

Thank you!

**Comment by JamesBlack2** [comment]
There two questions:

1：

               2 > Restoring packages of H: \Test\WpfApp6\WpfApp6. csproj...

In this step, I publish it to the local directory, which took me at least a few minutes. I tested several projects, all of then are took several minutes to compile the first time. It's just a small project. would bigger projects take longer?

2：In that console application, it can be run normally. but if it is WPF project,  it will crashes at the start to run.

**Comment by rladuca** [comment]
@JamesBlack2 It would be helpful if you could upload a sample project to a GitHub repository and then share that repository with us here.  Then we can examine the exact code that you are having issues with.

**Comment by JamesBlack2** [comment]
Thanks,i has update the sample project to github  now.
https://github.com/JamesBlack2/WpfApp6.git

**Comment by lindexi** [comment]
@JamesBlack2 I think he builds the application in the win10 and runs in win7.

**Comment by lindexi** [comment]
How to make the WPF run in dotnet core 3.0, See [如何创建一个基于 .NET Core 3 的 WPF 项目 - walterlv](https://walterlv.com/post/create-new-wpf-on-dotnet-core-project.html )

[将基于 .NET Framework 的 WPF 项目迁移到基于 .NET Core 3 - walterlv](https://walterlv.com/post/migrate-wpf-project-from-dotnet-framework-to-dotnet-core.html )

**Comment by JamesBlack2** [comment]
It is,build in Win10 and run in Win7.it must be build in Win7 if i want it run in win7 or other win version?
I saw that guide like that,and build to try run,the result is the same too.the project is make by command "dotnet new wpf -o WalterlvWpfApp"

**Comment by shadowaltar** [comment]
> 我使用VS2019新建了一个测试用的.net core的控制台应用程序，但是当我把这程序发布后生成的文件复制到Win7 service 1,7061的 32位系统想远行时，却提示文件版本与系统不兼容，就是将其设置兼容性再跑时，也不行。是否VS2019预览版并没有完全将.net core的运行打包在一起？
> 
> 下面是app的项目配置。
> 《Project Sdk="Microsoft.NET.Sdk">
> 《 PropertyGroup>
> 《 OutputType>Exe《/OutputType>
> 《 TargetFramework>netcoreapp2.2《/TargetFramework>
> 《/PropertyGroup>
> 《/Project>

I used VS2019 to create a .Net Core console app for testing: when I put the files built to Windows7 Service 1,7061 32bits, it prompts that the file version is incompatible with the system. Nor would it work if I turn on the compatibility mode.
Does it mean that VS2019 Preview doesn't bundle the .Net Core runtime?

@karelz here you are.

**Comment by boyjiwen** [comment]
I also encountered this problem.
Created a wpf program based on .net core3.0, which can run normally on this machine and other WIN10 machines, but it can't run under WIN7 (x64 and x86 try, the result is the same). Also try to install the .netcore3.0 runtime on the WIN7 machine, the result is still the same, the WPF window does not display and an error window pops up.
After the x86 app has replaced some of the api-ms-win-core-xxxx.dll, the WPF window is displayed, but the error window pops up, and the program exits after clicking.

**Comment by randyliang1** [comment]
我在win 7上使用 .net core3.0 发布wpf 工程，也遇到这个问题。

**Comment by LiangWeixin** [comment]
Me too, I created the .net core 3.0 program under win7, but it can't run. Is it that some of my configurations are wrong?

**Comment by lindexi** [comment]
@randyliang1 @LiangWeixin I can run it in win7 sp1 

![image](https://user-images.githubusercontent.com/16054566/49644124-87f45880-fa52-11e8-86ea-1d68794fc255.png)

See [如何创建一个基于 .NET Core 3 的 WPF 项目 - walterlv](https://walterlv.com/post/create-new-wpf-on-dotnet-core-project.html )

[将基于 .NET Framework 的 WPF 项目迁移到基于 .NET Core 3 - walterlv](https://walterlv.com/post/migrate-wpf-project-from-dotnet-framework-to-dotnet-core.html )

**Comment by JamesBlack2** [comment]
is you build you project in win7 ,and you run system has setup .netcore runtime? @lindexi

**Comment by boyjiwen** [comment]
> @randyliang1 @LiangWeixin I can run it in win7 sp1
> 
> ![image](https://user-images.githubusercontent.com/16054566/49644124-87f45880-fa52-11e8-86ea-1d68794fc255.png)
> 
> See [如何创建一个基于 .NET Core 3 的 WPF 项目 - walterlv](https://walterlv.com/post/create-new-wpf-on-dotnet-core-project.html)
> 
> [将基于 .NET Framework 的 WPF 项目迁移到基于 .NET Core 3 - walterlv](https://walterlv.com/post/migrate-wpf-project-from-dotnet-framework-to-dotnet-core.html)

Is the running environment locally installed with VS2019 or net core3 sdk?
After I compile the WIN7 device, the local can run normally, and I can't run it on another WIN7 machine without a development environment.

![image](https://user-images.githubusercontent.com/16402158/49649425-16250a80-fa64-11e8-95e3-12c51cd4e1c5.png)

**Comment by rladuca** [comment]
I can't tell what the actual error you're experiencing is.  However, WPF for .NET Core currently has a dependency on the VCRuntime (See the [getting started guide](https://github.com/dotnet/wpf/blob/master/Documentation/getting-started.md#known-issues) and #37).  The fact that various machine are behaving this way with a previously published application seems to be indicative of running into this requirement.

If you can verify that your issues are spawned from the same (or similar) issue and that installing VCRuntime on all machines that you want to run your application on solves the issue we can close this as a duplicate of #37.  We are already working to address that specific issue.

**Comment by boyjiwen** [comment]
@rladuca 
After installing VCRuntime x86 and x64 (Microsoft Visual C++ Redistributable for Visual Studio 2017) on the WIN7 machine, the problem still exists.

![image](https://user-images.githubusercontent.com/16402158/49660033-8f325b00-fa80-11e8-9db7-27ddbeb33460.png)

Please advise how can I find my program error? Have tools?

My requirement is that the program can run under a machine that does not have the .netcore runtime installed.

Also, will the final version support Windows XP?

**Comment by rladuca** [comment]
When I took the base project from `dotnet new wpf` and did an initial publish selecting win-x86 as the target, it did publish a correct executable, but the binary api-ms-win-crt-l1-1-9.dll (as stated above by @boyjiwen ) seems to be the incorrect bitness.

I'm not sure exactly what's going on right now, I'll have to dig deeper and get some help from people more familiar with the publish step in VS, but this seems like a bug, though not in WPF specifically.  Perhaps I am still doing something wrong myself.

What worked for me is to publish from the command line.  On your dev box, just do the following `dotnet publish -r win10-x86 --self-contained` in the directory where the solution is.  The output from this ran on my Win7 SP1 machine with no issues.

I appreciate everyone's help in overcoming any translation barrier and providing us with effective instructions to reproduce the problem!

**Comment by rladuca** [comment]
I was able to make this work in VS2019 by doing the following:

```xml
<Project Sdk="Microsoft.NET.Sdk.WindowsDesktop">

  <PropertyGroup>
    <OutputType>WinExe</OutputType>
    <TargetFramework>netcoreapp3.0</TargetFramework>
    <UseWPF>true</UseWPF>
    <SelfContained>true</SelfContained>
    <RuntimeIdentifier>win7-x86</RuntimeIdentifier>
  </PropertyGroup>

</Project>
```

When I published this from VS and ran the output on Win7 SP1 VM with only the VCRuntime2017 installed, the application loaded as expected.

I am getting some more guidance as to a sample I can point to that is more definitive, but if you try that you should be able to get it working.

**Comment by karelz** [comment]
@rladuca Supported OS versions in core repo: https://github.com/dotnet/core/blob/master/os-lifecycle-policy.md (3.0 is still TBD, but you can check 2.1)

**Comment by JamesBlack2** [comment]
![image](https://user-images.githubusercontent.com/10073006/49683250-6485e800-fafc-11e8-81ae-6503043beff6.png)
![image](https://user-images.githubusercontent.com/10073006/49683471-81bcb580-fb00-11e8-960e-e43a8c30f611.png)

I has got the dump file by TaskMgr and update to github,that may be help: https://github.com/JamesBlack2/WalterlvWpfApp.git


the app can pop the notice message box normally ,but when click ok later it crashed. 
the app published by this command.**dotnet publish -r win10-x86 --self-contained**,the vcruntimer has installed.

**Comment by lindexi** [comment]
@JamesBlack2 you should not publish in win10. I can not find any in your dump.

**Comment by boyjiwen** [comment]
@rladuca @karelz 

After Windows 7 installed VS2019, the WPF program that could not be run before can now run normally.

Now find a way to check if the WPF program depends on a package in VS2019.

**Comment by boyjiwen** [comment]
@rladuca @karelz 
Tested with 6 WIN7 machines, VC2017 Runtime can run on the machine.

Before the test machine installed VC2017 can not run (downloaded from Microsoft officially),
1. Normal after installing VS2019;
2. Cannot run after deleting VC2017 runtime, normal after reinstalling VC2017.
  There may be a problem with the runtime library downloaded from Microsoft.

The problem should be the same as #37. Will this problem be solved in the final version?
We plan to transfer the current WPF program to .net core3. When will the final version be released?

**Comment by lindexi** [comment]
@boyjiwen do not impatient to update WPF to dotnet core 3. I update some WPF application but I find some application be crash in running.

**Comment by boyjiwen** [comment]
@lindexi 
The .net framework45 installer will fail to install on some Win7 machines.
Therefore, our program has two runs of .net35 and .net45. Under WIN7, the launcher will launch that version in the system environment. But there are still many users who can't run our programs.
At the same time, we also need to provide users with no installation version.

I tried mono before, but only supported Winform. I wanted to extract the .net framework to the installation package before, but I didn't find the relevant information.

If this can solve the dependence on the .NET runtime under WIN7, DAU should be able to increase by 15%~20%.

Why not use C++/Qt/DirectUI or Html5 solutions?
In the past, we used C++/Qt solutions, but the development efficiency was low, the resource investment was not as much as mobile development, and it was too difficult to recruit experienced Windows C++ developers in China.

First we have UWP products and then ported to WPF. After using WPF, we found that development efficiency and product stability are better. At present, our WPF program reaches millions of daily active users.

Therefore, we are more eager and hope that the final stable version of .net core3 will be released soon.

**Comment by rladuca** [comment]
@boyjiwen, we intend to fix the vcruntime dependency prior to RTM.  We don't have a specific time frame for that otherwise.

WPF does not depend on VS2019 in any way at runtime.  I have repeated the publishing experiment multiple times on clean Win7 VMs and VCRuntime is the only dependency to a properly published WPF application.

@JamesBlack2 I haven't had a chance to check out the latest dump file you posted.  I will do so tomorrow.

**Comment by lindexi** [comment]
@boyjiwen Try to use [turbo studio](https://turbo.net/studio).

**Comment by boyjiwen** [comment]
@rladuca @lindexi 
Copy vcruntime140.dll to the released package, it can run normally on Win7 machine without VCRuntime installed.

**Comment by rladuca** [comment]
@boyjiwen Yes, that is expected.  The eventual solution for RTM is something relatively similar to that, but with some more considerations in mind.

For instance, the framework itself cannot carry vcruntime140.dll, because an application may have its own version of the DLL.  Therefore, the framework needs to have a renamed version of it, which complicates matters in linking.

Eventually, we will contain a private version of the DLL that will be brought along during a publish and not conflict with the application's use of its own VCRuntime (if it happens to use it).

For a specific application being published, you can definitely copy along VCRuntime as you can be 100% sure that it's the right thing to do.

**Comment by boyjiwen** [comment]
@rladuca 
Today builds a WPF version with vcrtime, tested internally, with another problem found on several machines:

The program can be run and the MessageBox is successfully called, but after confirming it, it crashes.

![image](https://user-images.githubusercontent.com/16402158/49867974-fe27ff00-fe46-11e8-96c0-205ec29ebdc6.png)
![image](https://user-images.githubusercontent.com/16402158/49867984-05e7a380-fe47-11e8-894f-78279458c95e.png)
![image](https://user-images.githubusercontent.com/16402158/49867991-0aac5780-fe47-11e8-9b62-11a6253b78be.png)

**Comment by boyjiwen** [comment]
@JamesBlack2  Your problem is encountered when I installed the virtual machine today, installed NetcoreRuntime, using the dotnet command.
Dotnet --list-runtimes => Failed to load the hostfxr.dll
After installing the system update patch KB2533623, you should be able to solve the problem.

At this time, I realized that the .net core may be related to the system update patch. After installing KB2533623, the program can display Messabox, but it will crash and display in KernalBase.dll.
I checked a few machines and found that the high version of KernalBase.dll works fine, but the lower version of KernalBase.dll will pop up Messabox and then crash.

After updating other system update patches (re-verification is the update patch), it can run normally.
At this time, the viewing system has improved the version of KernelBase.dll.

@rladuca @karelz @lindexi 
Is there any plan to reduce the reliance on system update patches?
A large part of the WIN7 user system update function is turned off. When the relevant update patch is not installed, the WPF runs incorrectly.

I hope that these dependencies can be reduced later. In this case, Windows software development has developed an efficient UI framework to improve development efficiency, which may attract more enterprises and developers to participate in Windows development.

**Comment by stevenbrix** [comment]
There is a lot of awesome discussion here, thank you everyone for contributing! I'm not sure it's possible to remove the dependency on those two KB updates, they deliver necessary components that are not currently present on the machine. I'm going to close this as a duplicate of #37, if that is incorrect please respond and we can continue the discussion.

@boyjiwen can you please file an issue regarding the message box crash (if one doesn't already exist)? Thanks!
